### PR TITLE
Fix #2686: deprecate allow_plain_identifiers option

### LINKED ISSUE #2686: [RFC] Deal with plain identifiers
Related:

https://github.com/api-platform/core/pull/1834
#2521 #1833 https://github.com/api-platform/core/pull/1716 

What we need to do:

https://github.com/api-platform/core/pull/2022#issuecomment-398851861 

This will help close every plain identifier issue.

Also needed with JsonApi see https://github.com/api-platform/core/issues/1833 and https://github.com/api-platform/core/pull/1834

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Branch?       | main
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Fix #2686 
| License       | MIT
| Doc PR        | api-platform/docs#1326

### CODE REVIEW

File: src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 1]
```diff
@@ -177,6 +177,7 @@ private function registerCommonConfiguration(ContainerBuilder $container, array
         $container->setParameter('api_platform.formats', $formats);
         $container->setParameter('api_platform.patch_formats', $patchFormats);
         $container->setParameter('api_platform.error_formats', $errorFormats);
+        // todo to remove in 3.0
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php:180]:
```suggestion
        // TODO: to remove in 3.0
```

File: tests/Serializer/AbstractItemNormalizerTest.php [hunk 2]
```diff
@@ -942,6 +942,9 @@ public function testChildInheritedProperty(): void
         ], $normalizer->normalize($dummy, null, ['resource_class' => DummyTableInheritance::class, 'resources' => []]));
     }
 
+    /**
+     * todo to remove in 3.0
```
Comment (soyuka) [inline tests/Serializer/AbstractItemNormalizerTest.php:946]:
```suggestion
     * TODO: to remove in 3.0
```

File: tests/Serializer/AbstractItemNormalizerTest.php [hunk 3]
```diff
@@ -995,6 +998,9 @@ public function testDenormalizeRelationWithPlainId()
         $propertyAccessorProphecy->setValue($actual, 'relatedDummy', $relatedDummy)->shouldHaveBeenCalled();
     }
 
+    /**
+     * todo to remove in 3.0
```
Comment (soyuka) [inline tests/Serializer/AbstractItemNormalizerTest.php:1002]:
```suggestion
     * TODO: to remove in 3.0
```

File: tests/Serializer/AbstractItemNormalizerTest.php [hunk 4]
```diff
@@ -1054,6 +1060,9 @@ public function testDenormalizeRelationWithPlainIdNotFound()
         $normalizer->denormalize($data, Dummy::class, 'jsonld');
     }
 
+    /**
+     * todo to remove in 3.0
```
Comment (soyuka) [inline tests/Serializer/AbstractItemNormalizerTest.php:1064]:
```suggestion
     * TODO: to remove in 3.0
```

File: tests/Fixtures/app/config/config_common.yml [hunk 5]
```diff
@@ -25,6 +25,7 @@ api_platform:
     description: |
         This is a test API.
         Made with love
+    # todo to remove in 3.0
```
Comment (soyuka) [inline tests/Fixtures/app/config/config_common.yml:28]:
```suggestion
     * TODO: to remove in 3.0
```

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php [hunk 6]
```diff
@@ -897,6 +897,7 @@ private function getPartialContainerBuilderProphecy($configuration = null)
             'api_platform.title' => 'title',
             'api_platform.version' => 'version',
             'api_platform.show_webby' => true,
+            // todo to remove in 3.0
```
Comment (soyuka) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php:900]:
```suggestion
     * TODO: to remove in 3.0
```

File: src/Serializer/AbstractItemNormalizer.php [hunk 7]
```diff
@@ -851,6 +856,10 @@ private function setValue($object, string $attributeName, $value)
         }
     }
 
+    /**
+     * todo to remove in 3.0
```
Comment (soyuka) [inline src/Serializer/AbstractItemNormalizer.php:860]:
```suggestion
     * TODO: to remove in 3.0
```

File: src/Serializer/AbstractItemNormalizer.php [hunk 8]
```diff
@@ -88,7 +88,12 @@ public function __construct(PropertyNameCollectionFactoryInterface $propertyName
         $this->resourceClassResolver = $resourceClassResolver;
         $this->propertyAccessor = $propertyAccessor ?: PropertyAccess::createPropertyAccessor();
         $this->itemDataProvider = $itemDataProvider;
+
+        if (true === $this->allowPlainIdentifiers) {
+            @trigger_error(sprintf('Allowing plain identifiers as argument of "%s" is deprecated since API Platform 2.7 and will not be possible anymore in API Platform 3.', self::class), \E_USER_DEPRECATED);
```
Comment (soyuka) [inline src/Serializer/AbstractItemNormalizer.php:93]: do you think we could try and run our tests without plain identifiers to see what would potentially break ?

File: src/Bridge/Symfony/Bundle/Resources/config/graphql.xml [hunk 9]
```diff
@@ -216,6 +216,7 @@
             <argument type="service" id="api_platform.name_converter" on-invalid="ignore" />
             <argument type="service" id="serializer.mapping.class_metadata_factory" on-invalid="ignore" />
             <argument type="service" id="api_platform.item_data_provider" on-invalid="ignore" />
+            <!-- todo to remove in 3.0 -->
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/Resources/config/graphql.xml:219]:
```suggestion
            <!-- TODO: to remove in 3.0 -->
```

File: src/Bridge/Symfony/Bundle/Resources/config/api.xml [hunk 10]
```diff
@@ -111,6 +111,7 @@
             <argument type="service" id="api_platform.name_converter" on-invalid="ignore" />
             <argument type="service" id="serializer.mapping.class_metadata_factory" on-invalid="ignore" />
             <argument type="service" id="api_platform.item_data_provider" on-invalid="ignore" />
+            <!-- todo to remove in 3.0 -->
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/Resources/config/api.xml:114]:
```suggestion
            <!-- TODO: to remove in 3.0 -->
```

**Review by soyuka — COMMENTED** [review]
Not sure about the `todo` => `TODO:` format ^^

File: src/Serializer/AbstractItemNormalizer.php [hunk 11]
```diff
@@ -88,7 +88,12 @@ public function __construct(PropertyNameCollectionFactoryInterface $propertyName
         $this->resourceClassResolver = $resourceClassResolver;
         $this->propertyAccessor = $propertyAccessor ?: PropertyAccess::createPropertyAccessor();
         $this->itemDataProvider = $itemDataProvider;
+
+        if (true === $this->allowPlainIdentifiers) {
```
Comment (guilliamxavier) [inline src/Serializer/AbstractItemNormalizer.php:92]: Bug
```suggestion
        if (true === $allowPlainIdentifiers) {
```

File: src/Serializer/AbstractItemNormalizer.php (hunk 8 above)
Comment (vincentchalamon) [reply src/Serializer/AbstractItemNormalizer.php:93]: I tried turning `allow_plain_identifiers` to false and run Behat tests, the only failing scenario are: features/json/relation.feature:154 & 187. I added a `@deprecated` tag and a TODO on them

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -95,7 +95,11 @@ public function getConfigTreeBuilder()
                 ->scalarNode('name_converter')->defaultNull()->info('Specify a name converter to use.')->end()
                 ->scalarNode('asset_package')->defaultNull()->info('Specify an asset package name to use.')->end()
                 ->scalarNode('path_segment_name_generator')->defaultValue('api_platform.path_segment_name_generator.underscore')->info('Specify a path name generator to use.')->end()
-                ->booleanNode('allow_plain_identifiers')->defaultFalse()->info('Allow plain identifiers, for example "id" instead of "@id" when denormalizing a relation.')->end()
+                ->booleanNode('allow_plain_identifiers')
+                    ->defaultFalse()
+                    ->info('Allow plain identifiers, for example "id" instead of "@id" when denormalizing a relation.')
+                    ->setDeprecated(...$this->buildDeprecationArgs('2.7', 'The use of the `allow_plain_identifiers` has been deprecated in 2.7 and will be removed in 3.0.'))
```
Comment (alanpoulain) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php:101]:
```suggestion
                    ->setDeprecated(...$this->buildDeprecationArgs('2.7', 'The use of `allow_plain_identifiers` has been deprecated in 2.7 and will be removed in 3.0.'))
```

**Review by alanpoulain — APPROVED** [review]

**Comment by soyuka** [comment]
thanks @vincentchalamon

**Comment by dunglas** [comment]
In my opinion we should still provide support for plain identifiers but as a normalizer/de normalizer decorator (that will simplify everything).

**Comment by vincentchalamon** [comment]
@dunglas that's the goal of this PR in the documentation :wink:: https://github.com/api-platform/docs/pull/1326

**Pull request merged**